lib/grammers-client/src/client/files.rs

```diff
@@ -325,6 +335,70 @@ impl Client {
         // files, so it's fine to unwrap.
         let name = path.file_name().unwrap().to_string_lossy().to_string();
 
-        self.upload_stream(&mut file, size, name).await
+        self.upload_stream(file, size, name).await
+    }
+}
+
+struct PartStreamInner<S: AsyncRead + Unpin + Send + 'static> {
+    stream: S,
+    current_part: i32,
+}
+
+struct PartStream<S: AsyncRead + Unpin + Send + 'static> {
+    inner: AsyncMutex<PartStreamInner<S>>,
+    total_parts: i32,
+}
+
+impl<S: AsyncRead + Unpin + Send> PartStream<S> {
+    fn new(stream: S, size: usize) -> Self {
+        let total_parts = ((size + MAX_CHUNK_SIZE as usize - 1) / MAX_CHUNK_SIZE as usize) as i32;
+        Self {
+            inner: AsyncMutex::new(
+                "upload_stream",
+                PartStreamInner {
+                    stream,
+                    current_part: 0,
+                },
+            ),
+            total_parts,
+        }
+    }
+
+    fn total_parts(&self) -> i32 {
+        self.total_parts
+    }
+
+    async fn next_part(&self) -> Result<Option<(i32, Vec<u8>)>, io::Error> {
+        let mut lock = self.inner.lock("read part").await;
+        if lock.current_part >= self.total_parts {
```

Comment:
Not what I had in mind w.r.t. getting rid of the atomic but fair enough. I presume this is a bit safer than having the parts without a lock.

Reply:
Since the part count is always used when the lock is held, there should be no problem here. Also mutating directly the part count is unsafe and therefore must be behind some kind of mutex.